presto-main/src/main/java/com/facebook/presto/sql/planner/DistributedExecutionPlanner.java

```diff
@@ -389,4 +409,61 @@ private Visitor(Session session)
             throw new UnsupportedOperationException("not yet implemented: " + node.getClass().getName());
         }
     }
+
+    private static class DynamicFilterDescriptionFutureSupplier
```

Comment:
Just return future directly from the service. Listener is supposed to be constantly listening for the events. Here just future is more natural. Because you are in fact need to wait for just a deferred value.